recipe/meta.yaml

```diff
@@ -42,14 +40,15 @@ requirements:
     - python
     - pip
     - setuptools
-    - llvmlite 0.38.*
+    - llvmlite 0.39.*
     - numpy
-    - tbb-devel 2021.*
+    - tbb-devel 2021.*,<2021.6
 
   run:
     - python
     - {{ pin_compatible('llvmlite', max_pin='x.x') }}
     - {{ pin_compatible('numpy') }}
+    - importlib-metadata  # [py < 39]
     # needed for pkg_resources
     - setuptools
```

Comment:
Noticed `setuptools` is [constrained to `<60` in requirements]( https://github.com/numba/numba/blob/d6731f6d22f6695ee7abc305f81f81115639e2a4/buildscripts/condarecipe.local/meta.yaml#L49 ). Why was this needed?

Reply:
Originally we noticed that there was a problem with setuptools 65 (Numba issue: https://github.com/numba/numba/issues/8355). NumPy produced a deprecation warning stating:

```
  `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
  of the deprecation of `distutils` itself. It will be removed for
  Python >= 3.12. For older Python versions it will remain present.
  It is recommended to use `setuptools < 60.0` for those Python versions.
  For more details, see:
    https://numpy.org/devdocs/reference/distutils_status_migration.html 
```

This is why we constrained it to <60. However, this does cause other issues, as noted in https://github.com/numba/numba/pull/8366#issuecomment-1258115186.

The plan is to make setuptools an optional runtime dependency in Numba, and not pin to any specific version (see https://github.com/numba/numba/pull/8366#issuecomment-1259804254) - so I think it will be OK not to constrain to any specific setuptools version, as long as the packages build successfully.

Reply:
Gotcha. Is `pycc` using `numpy.distutils` or just `distutils` (and `setuptools` ofc)?

Reply:
It seems to be importing things from both `numpy.distutils` and `distutils`.

Reply:
There's a plan to remove this dependency and constraint outlined here: https://github.com/numba/numba/pull/8366#issuecomment-1259804254 which follows on from various discussions yesterday. I don't think this is relevant to the current release as it will require additional code changes which are likely to be in 0.57 (next release).

Reply:
Hmm. The challenge is constraining `setuptools` in this way in a user environment would be quite limiting and likely cause different kinds of pain. Given this, am wondering if it make sense to just leave that constraint out (at least for `run`). The user would at least get the deprecation warning above if they used an incompatible version of `setuptools` with `pycc`

Reply:
I think use of `setuptools > 60` will potentially trigger a warning from NumPy which could potentially make at least one of Numba's unit tests that assert e.g. no warnings on import fail. There's also a version of setuptools 0.65.x that broke Numba (and NumPy). This lead to the constraint above being added based on the recommendation from NumPy (see below). Given it's only the `pycc` part of Numba that uses distutils/setuptools and the vast majority of users will not be using `pycc`, it _might_ be ok to leave out the constraint at run time, but it would seemingly go against NumPy's recommendation.
```
$ python -c "import numpy; print(numpy.__version__); import numpy.distutils.misc_util"
1.23.1
<string>:1: DeprecationWarning: 

  `numpy.distutils` is deprecated since NumPy 1.23.0, as a result
  of the deprecation of `distutils` itself. It will be removed for
  Python >= 3.12. For older Python versions it will remain present.
  It is recommended to use `setuptools < 60.0` for those Python versions.
  For more details, see:
    https://numpy.org/devdocs/reference/distutils_status_migration.html
```

Reply:
I think it's OK to leave the constraint out - if it is left out, then I think the worst that will generally happen is that users might see a deprecation warning from NumPy. There were some setuptools 65 versions which were completely broken for Numba (65.0.0 and 65.0.1) that were fixed in 65.0.2: https://setuptools.pypa.io/en/latest/history.html#v65-0-2 - so I don't think it's necessary to constrain `setuptools` here.

Reply:
Limiting setuptools is not feasible and should only be done in extreme cases - where things actually break. This is very much not an extreme case. NumPy is making that recommendation because they are removing this and don't want to touch it again. IIRC, users can also set `SETUPTOOLS_USE_DISTUTILS=stdlib` to make NumPy and setuptools>=60 work for the moment.

If something bad happens, we can patch in a constraint after the fact if we have to.

Reply:
Thank you all for the input, it sounds like the consensus is "don't constrain the `setuptools` version" and there's a way to fix it if there is a problem.

Reply:
Sounds good. Thanks all for your feedback 🙏

Reply:
As we are diverging here now from the official package constraints, this sadly causes all packages that depend on `numba` to have a failing `pip check`. I would argue that we should align with upstream here. If `numba` (maintainers) thinks this is not an issue, then it should be removed upstream (as done in one of the linked PRs which probably takes a bit longer). As long as upstream defines this constraint and it is in the `pip` metadata, we should reflect that in the conda metadata. I went ahead and made a PR to address that: https://github.com/conda-forge/numba-feedstock/pull/104

Reply:
I do see it [here](https://github.com/numba/numba/blob/061f21a73a574ea1b3451910a36db4f9b3ae788b/setup.py#L378). 😱

Reply:
> Why does our pip check pass but not others?

Probably because there is a `run_test.sh` the entries in `test: commands: …` are not executed. I don't see it in the logs.

Reply:
Numba version 0.56.3 has been released with no restrictions on the `setuptools` version so as to address this issue. Thanks for the feedback.